### PR TITLE
Feature/group api

### LINKED ISSUE #88: Chaincode - Group-Contract
Need to re-invent Hyperledgers Group-System to ensure fast implementation / proof of concept of group-checking within the smart contracts.

### PR DESCRIPTION
**Pull request by anemonekk**

### CODE REVIEW

**Review by mco33415 — APPROVED** [review]

**Review by matthias-geuchen — APPROVED** [review]

File: hlf/chaincode/group.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+## Transactions
+
+### Add User To Group
+- ID = addUserToGroup
+- Send
+    - enrollmentID :: String
+    - groupID :: String
+- Receive
+    - userList :: List\<String\>
+        - Description: ledgerState after transaction.
```
Comment (anemonekk) [inline hlf/chaincode/group.md]:
```suggestion
- Receive
    - group :: [Group](#Group)
        - Description: ledgerState after transaction.
```

**Review by NikoBergemann — APPROVED** [review]

File: hlf/chaincode/group.md [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+## Transactions
+
+### Add User To Group
+- ID = addUserToGroup
+- Send
+    - enrollmentID :: String
+    - groupID :: String
+- Receive
+    - group :: [Group](#Group)
+        - Description: ledgerState after transaction.
```
Comment (dhelmert) [inline hlf/chaincode/group.md]:
```suggestion
    - ""
      -  Description: Done; returns empty String on success.
```

File: hlf/chaincode/group.md [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+## ContractName "UC4.Group"
+
+## Transactions
+
+### Add User To Group
+- ID = addUserToGroup
+- Send
+    - enrollmentID :: String
+    - groupID :: String
```
Comment (dhelmert) [inline hlf/chaincode/group.md]:
```suggestion
    - enrollmentId :: String
    - groupId :: String
```

File: hlf/chaincode/group.md [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+      }
+      ```
+      - Description: This error is returned, if the required approvals are not present.
+
+
+### Remove From Group
+- ID = removeUserFromGroup
+- Send
+    - enrollmentID :: String
+    - groupID :: String
```
Comment (dhelmert) [inline hlf/chaincode/group.md]:
```suggestion
    - enrollmentId :: String
    - groupId :: String
```

File: hlf/chaincode/group.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+- Receive
+    - userList :: List\<String\>
+            - Description: ledgerState after transaction.
```
Comment (dhelmert) [inline hlf/chaincode/group.md]:
```suggestion
    - ""
      -  Description: Done; returns empty String on success.
```

File: hlf/chaincode/group.md [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+      For detailed informations see [parameter checks](#parameterChecks).
+
+### Get Groups For User
+- ID = getGroupsForUser
+- Send
+    - enrollmentId :: String
+
+- Receive
+    - groupList :: List\<String\>
+          - Description: Exhaustive List of all group names the user is part of.
```
Comment (dhelmert) [inline hlf/chaincode/group.md]:
```suggestion
          - Description: Exhaustive List of all groupIds the user is part of.
```

File: hlf/scala/group.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+Any Method can throw these [Exceptions](errors.md#Exceptions).
+
+For "group" handling we offer the following interface. 
+Additionally, as described in [General Communication](general-communication.md), different proposal calls can be made.
+
+
+## AddUserToGroup(enrollmentId: String, groupId: String)
+- Returns
+    - userList :: Json (List:\<String\>)
+        - => Success
```
Comment (dhelmert) [inline hlf/scala/group.md]:
```suggestion
    - ""
    => Success
```

File: hlf/scala/group.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+          - If the given parameters could not be parsed.
+          - If the given transaction could not be performed due to semantic reasons.
+    - TransactionError :: Json (refer to: [GenericError](../chaincode/errors.md#GenericError))
+        - => error is returned, 
+          - If the required approvals are not present on the ledger.
+
+## RemoveUserFromGroup(enrollmentId: String, groupId: String)
+- Returns
+    - userList :: Json (List:\<String\>)
+        - => Success
```
Comment (dhelmert) [inline hlf/scala/group.md]:
```suggestion
    - ""
    => Success
```

File: hlf/scala/group.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+- Returns
+    - groupList :: List\<[Group](../chaincode/group.md#Group)\>
+    => Success
+
+> Note! Invalid States on the ledger get ignored.
+> We return all valid Groups that match the filters.
+
+## GetUsersForGroup(groupId: String)
+- Returns
+    - groupList :: List\<String\>
```
Comment (dhelmert) [inline hlf/scala/group.md]:
```suggestion
    - userList :: List\<String\>
```

File: hlf/scala/group.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+          - If the required approvals are not present on the ledger.
+          - If the state of data on the ledger is not consistent with the curent model.
+            - This error should only occurr if the model changes while the old ledger state remains without modification.
+    - TransactionError :: Json (refer to: [DetailedError](../chaincode/errors.md#DetailedError))
+        - => error is returned
+          - If the given parameters could not be parsed.
+
+## GetGroupsForUser(enrollmentId: String)
+- Returns
+    - groupList :: List\<String\>
```
Comment (dhelmert) [inline hlf/scala/group.md]:
```suggestion
    - groupIdList :: List\<String\>
```

File: hlf/chaincode/group.md [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+      - Description: This error is returned, if the given parameters are not well formatted, they are listed in invalidParams.  
+      For detailed informations see [parameter checks](#parameterChecks).
+
+### Get Groups For User
+- ID = getGroupsForUser
+- Send
+    - enrollmentId :: String
+
+- Receive
+    - groupList :: List\<String\>
```
Comment (dhelmert) [inline hlf/chaincode/group.md]:
```suggestion
    - groupIdList :: List\<String\>
```

**Review by dhelmert — CHANGES_REQUESTED** [review]

File: hlf/scala/group.md (hunk 9 above)
Comment (RBeierling) [reply hlf/scala/group.md]: resolve this and get my approval

File: hlf/chaincode/group.md [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+      ```
+      - Description: This error is returned, if the required approvals are not present.
+
+### Remove User From All Groups
+- ID = removeUserFromAllGroups
+- Send
+    - enrollmentId :: String
+- Receive
+    - ""
+      -  Description: Done, returns the submitted data.
```
Comment (dhelmert) [inline hlf/chaincode/group.md]:
```suggestion
      -  Description: Done; returns empty String on success.
```

File: hlf/chaincode/group.md [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+      - Description: This error is returned, if the given parameters are not well formatted, they are listed in invalidParams. It is also returned, if the user is not registered.  
+      For detailed informations see [parameter checks](#parameterChecks) and [semanticChecks](#semanticChecks).
+
+## <a id="Models" />Models
+
+### <a id="Group" />Group
+```json
+{
+  "groupId": "0123456ExampleGroup",
+  "userList": ["0123456", "123"]
```
Comment (dhelmert) [inline hlf/chaincode/group.md]:
```suggestion
  "userList": ["EnrollmentIdExample123", "EnrollmentIdExample124"]
```

**Review by dhelmert — CHANGES_REQUESTED** [review]

**Review by RBeierling — APPROVED** [review]

**Review by dhelmert — APPROVED** [review]

**Pull request merged**